### ISSUE TITLE
[vLLM] Per-user last-token gather index computed via bf16 matmul → wrong logits for batched users past flat-index 512

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-xla#5116]
## Problem

Greedy (`temperature=0`) decoding of **N identical prompts in one batch** does not produce identical outputs across all rows. Each row in a batch is an independent computation, so argmax decoding must be identical for every row — divergence is a correctness bug. First spotted in `tests/benchmark/test_vllm_benchmarks.py::test_vllm_benchmark[llama-3.2-3b]` locally and on Nightly CI (https://github.com/tenstorrent/tt-xla/actions/runs/27175829468/job/80224459033#step:15:123), reproducible using simpler case below.

The divergence has a clean, structural signature: the batch splits into **contiguous groups of rows** that each compute a different result (e.g. rows `0..k-1` emit one token, rows `k..N-1` emit another). It is **deterministic**, and it is **fully eliminated by enabling fp32 destination accumulation** (`fp32_dest_acc_en=True`). It is therefore a **bf16 destination-accumulation** problem in the batched prefill graph (tt-mlir compute-kernel-config family).

## Reproduction

Llama-3.2-3B-Instruct, 32 identical prompts, greedy, generating a single token (pure prefill):

```python
import vllm

PROMPT = "Here is an exhaustive list of the best practices for writing clean code:"
N = 32

llm = vllm.LLM(
    model="meta-llama/Llama-3.2-3B-Instruct",
    max_model_len=128,
    max_num_seqs=N,
    max_num_batched_tokens=N * 128,        # prefill all N prompts together
    gpu_memory_utilization=0.5,
    enable_prefix_caching=False,
    additional_config={"fp32_dest_acc_en": False},
)
sp = vllm.SamplingParams(temperature=0.0, max_tokens=1, ignore_eos=True)
outs = llm.generate([PROMPT] * N, sp)

ids = [o.outputs[0].token_ids[0] for o in outs]
for i, t in enumerate(ids):
    print(f"row {i:2d}: token {t}")
print("unique tokens:", len(set(ids)))   # bug: 2 (should be 1)
```

Observed (deterministic): rows 0–3 emit token `720`, rows 4–31 emit token `320`; `unique tokens = 2`. With `fp32_dest_acc_en=True` all 32 rows emit `720` (`unique tokens = 1`). Full captured output: https://gist.github.com/kmabeeTT/dd7524de473cd8158040fc860886cafd

Reproduced at tt-xla `8a54156b5` (`main`) with stock tt-mlir `c5f398432` (the version `main` pins), single device.

## Observations (to guide debugging)

1. **Contiguous-group split, boundary tracks the prefill shape.** The batch splits into a contiguous prefix `0..k-1` vs the rest `k..N-1`. Where the boundary `k` falls depends on the prefill shape (the program-config / core-grid the batched ops are given), **not** on a fixed user count:

   | additional_config | first divergent row |
   |:--|:--|
   | `{fp32_dest_acc_en: False}` | **4** (group 0–3 vs 4–31) |
   | `{fp32_dest_acc_en: False, min_context_len: 32}` | **16** (group 0–15 vs 16–31) |

   Forcing a 32-token minimum prefill context moves the boundary from 4 to 16. This is consistent with the cause being the **batch grouping of the runtime-selected program-config**, not a magic number. (Batches at or below the boundary are clean: with the boundary at 16, batch 4/8/16 all agree and batch 24 splits 16/8.)

2. **Deterministic.** Re-running an identical config gives byte-identical per-row output every time (same split, same token ids). Not a race.

3. **Layer-count independent.** Same behavior on a single-decoder-layer truncation (`additional_config={"num_hidden_layers": 1}`) and on the full 28-layer model; only the specific token ids differ.

4. **Not the sampler.** Forcing host-side sampling (`additional_config={"cpu_sampling": True}`) reproduces the split, so the wrong values are already in the **logits coming off the device**, not in on-device sampling/argmax.

5. **`fp32_dest_acc_en=True` eliminates it** — all rows agree. So the divergence is entirely a low-precision (bf16) destination-accumulation effect. (Note: enabling it on 8B-class models has been associated with a separate DRAM-OOM regression, so it is not a free global fix.)

6. **No single op reproduces it in isolation — even config-matched.** A standalone `ttnn` program running each production prefill op (the QKV / O / gate-up / down matmuls, dense causal SDPA, rms_norm) at the exact batched shapes with **N identical rows** produces **bit-identical** outputs across all rows — including when passed the *exact* compute config the lowered graph specifies (`math_fidelity=hifi4, fp32_dest_acc_en=False`). A row-local op (matmul / elementwise / norm) cannot, by construction, turn identical input rows into a grouped split. So the split is **introduced by graph composition**, not by any single kernel.

7. **Layouts are not the asymmetry.** All intermediate tensors in the lowered prefill graph are **interleaved DRAM** (no sharded layouts), and the matmul ops carry only `compute_config = hifi4` with **no explicit program_config** — the core grid is selected from the activation shape at runtime.

## Likely root cause

The combination of (5)+(6)+(7) points at the **in-graph program-config / core-grid that tt-metal selects for the batched prefill ops under the compiler-emitted `hifi4, fp32_dest_acc_en=False` config**: it partitions the batch into contiguous groups whose bf16 destination-accumulation reductions round differently, so one group's logits diverge enough to flip the greedy token. The boundary moving with the prefill shape (4 vs 16 above) is direct evidence of this grouping. fp32 destination accumulation has enough headroom that the groups agree.

This is the same family as the known compute-kernel-config issues:
- [tt-mlir #8666](https://github.com/tenstorrent/tt-mlir/issues/8666) — `fp32_dest_acc_en` not applied properly for `ttnn.matmul`.
- [tt-mlir #8657](https://github.com/tenstorrent/tt-mlir/issues/8657) — SDPA has no compute-kernel-config / fp32-accumulation path (`ttnn.scaled_dot_product_attention*` don't implement `TTNNComputeKernelConfigOpInterface`).

## Suggested next step

Localizing the exact op needs an **in-graph per-op activation dump** (tt-mlir runtime op callback / `dumpTensor`): run the batched prefill, dump every op's output, and find the first op whose output for a row past the boundary diverges from row 0 for identical inputs. That op's emitted program-config / core-grid is the one introducing the grouping asymmetry.

## Also visible in the existing benchmark

The same divergence surfaces in the existing `tests/benchmark/test_vllm_benchmarks.py -k llama-3.2-3b` run (batch 32, `fp32_dest_acc_en=False`), which prints multiple distinct completions for 32 identical prompts but still passes. That path generates 128 tokens, so it shows the divergence in a noisier form (decode mixed in); use the single-token snippet above to see the clean structural signature.

It is already live in nightly CI — e.g. the `perf vllm_llama_3_2_3b (n150-perf)` job at https://github.com/tenstorrent/tt-xla/actions/runs/27064401476/job/79883258025#step:14:121 (2026-06-06) printed 3 distinct completions for the 32 identical prompts and was still marked **success**.

## Why it isn't caught today

Benchmarks assert token counts and PCC of **User 0** only, so per-user divergence (Users 1..N) is invisible — a batch of 32 identical prompts can print 2 distinct completions and still pass.

**Comment by kmabeeTT** [comment]
@mmilosevicTT can you please take a look? I think this is in your wheelhouse?

**Comment by kmabeeTT** [comment]
(This came up in model-productization meeting today, need to tag as p0 (correctness issue, blocks chunked-prefill feature), claude chewed on it overnight and appears to have found the issue and gave me a local fix which I'm going to be using very short term (still need official fix to make it to tt-xla)...  talked to @mmilosevicTT offline about it - I am requesting that someone from compiler team take a closer look here please.

**Update: root-caused + candidate tt-mlir fix**

Dug in with a per-op activation dump (instrumented the tt-mlir runtime op loop to compare each op's per-user output for 32 identical prompts). New learnings:

- **The decoder backbone is bit-row-invariant** — all `[1024,…]` matmuls, SDPA, and rms_norm produce identical per-user output. So this is **not** a model-compute precision issue (correcting the original "matmul dest-accumulation in the model" framing).
- **First divergence is at the per-user last-token gather** feeding the LM head. The SHLO is a multi-dim `stablehlo.gather`; tt-mlir lowers it to `flat_index = indices @ strides` via a **float `ttnn.matmul` with bf16 destination accumulation** + embedding lookup.
- bf16 represents integers exactly only to 256 (step 2 to 512, **step 4 above 512**). So any user whose last-token **flat index ≥ 512** rounds to the wrong value → gathers the **wrong row** → wrong logits → wrong token. This exactly explains the boundary (first user hitting index 512: **user 16** at stride 32, **user 4** at stride 128), the determinism, and why global `fp32_dest_acc_en=True` masked it.

**Candidate fix (tt-mlir):** force `fp32_dest_acc_en` on matmuls whose result feeds an embedding's index operand (index arithmetic must be exact; these matmuls are tiny). Branch `kmabee/issue-5116-gather-index-fp32-dest-acc`, in `TTNNSetComputeKernelConfig` commit https://github.com/tenstorrent/tt-mlir/commit/164f371b06058075c62a216bb0197047745a1042

**Testing (Llama-3.2-3B, 32 identical prompts, greedy, global fp32 OFF):** all now consistent + correct —

| case | before | after |
|---|---|---|
| batch 32, 1-layer / full 28-layer | split | unique=1 |
| minimal config (boundary 4) | split | unique=1 |
| decode (16 / 32 tokens) | divergent | unique=1 |
| batch 1 | — | no regression |

Output matches the fp32 reference (correct, not just consistent). This is squarely the https://github.com/tenstorrent/tt-mlir/issues/8666 family but specifically the **gather/embedding index** path.


## Claude Debug Summary
Edit Jun 26: Here is summary of steps Claude used in degging this in a single session without any help: https://gist.github.com/kmabeeTT/d766b59a4521d680b4dcf4658f1f086b

**Comment by mmilosevicTT** [comment]
Reassinging to @sdjordjevicTT as this should be investigated by MLIR core team. The issue we experienced in training is not really related to this error, apart from using `fp32_dest_acc_en`.

**Comment by sdjordjevicTT** [comment]
Is this really a p0? This doesn't seem like a regression, if I am following the thread correctly?

**Comment by kmabeeTT** [comment]
@sdjordjevicTT  - Not a regression but pcc/correctness issue that affects all the models I've tried, so for correctness (we are trying to release model in tt-shield) it needs to be fixed I believe ...

**Comment by sdjordjevicTT** [comment]
Yes, I agree we need to fix it in the short term, but P1 seems like a much more reasonable priority to me.  :)

**Comment by sdjordjevicTT** [comment]
@nvukobratTT @dgolubovicTT, this seems to be some matmul precision issue. Can you guys pick this up and take a look at it? :)

**Comment by kmabeeTT** [comment]
Any chance to look closer at this one @dgolubovicTT ?  I've been using the cherry pick in tt-mlir I put up above locally, but would like to drop it for a real fix of some kind that can benefit released models.

**Comment by kmabeeTT** [comment]
## Extra Clarification + recent nightly note

**Which matmul (and why it's not greedy-only).** The offending op is **not** the LM-head matmul or the sampler — it's the small index matmul tt-mlir emits when lowering the per-user last-token `stablehlo.gather` (`select_hidden_states`: `hidden_states[batch_indices, logits_indices, :]`), which lowers to `typecast(i32→f32) → matmul(indices, strides) → ttnn.embedding`. The bf16-accumulating matmul rounds `flat_index = user*stride + offset` wrong above 512 → wrong gathered row → wrong logits. That's **upstream of both argmax and `ttnn.sample`**, so non-greedy is equally affected; greedy is just the detection lens. (Supersedes the body's "Likely root cause" — the backbone is bit-row-invariant per the per-op dump.)

**Mismatch stopped showing in nightly — likely masked, not fixed (unconfirmed).** Bisected the `perf llama_3_2_3b_instruct (n150-perf)` job:

- **Jun 22** (last split): tt-xla `9a570e41d499` — https://github.com/tenstorrent/tt-xla/actions/runs/27993693487/job/82932506260#step:15:134
- **Jun 23** (first all-matching): tt-xla `93ac846371fd` — https://github.com/tenstorrent/tt-xla/actions/runs/28066486199/job/83214437985#step:15:559

The only lowering-pipeline change in that window is **#5327 (default opt level → 1)**. Theory (unconfirmed): opt1 passes change the gather lowering / compute config so the float index matmul no longer misfires — i.e. it **masks** the latent bf16 lowering rather than fixing it. The targeted tt-mlir fix is still worth landing, perhaps.

**Comment by dgolubovicTT** [comment]
Here is the [PR](https://github.com/tenstorrent/tt-xla/pull/5599) that closes this issue. Everything is explained there. 
In short, we had exlicit default False valude for fp32_dest_acc_en in benchmark tests, which is wrong since compiler handles default setting of all compute kernel config attributes. More detailed explanation in PR description...
